src/allocators/HybridAllocator.sol

```diff
@@ -0,0 +1,230 @@
+        uint256[2][] calldata, /*idsAndAmounts*/ // The allocated token IDs and amounts.
+        bytes calldata allocatorData // Arbitrary data provided by the arbiter.
+    ) external view virtual returns (bool) {
+        if (claims[claimHash]) {
+            return true;
+        }
+
+        // Check the allocator data for a valid signature by an authorized allocator address
+        bytes32 digest = keccak256(abi.encodePacked(bytes2(0x1901), _COMPACT_DOMAIN_SEPARATOR, claimHash));
+        return _checkSignature(digest, allocatorData);
+    }
+
+    function _actualIdsAndAmounts(uint256[2][] memory idsAndAmounts) internal returns (uint256[2][] memory) {
+        uint256 idIndex = 0;
+        uint256 idsLength = idsAndAmounts.length;
+        if (idsLength == 0) {
+            revert InvalidIds();
+        }
+
+        // Check for native token - Native tokens must always be the first id
+        if (_splitToken(idsAndAmounts[0][0]) == address(0)) {
+            // Check allocator id
+            if (_splitAllocatorId(idsAndAmounts[0][0]) != ALLOCATOR_ID) {
+                revert InvalidAllocatorId(_splitAllocatorId(idsAndAmounts[0][0]), ALLOCATOR_ID);
+            }
+            if (idsAndAmounts[0][1] != 0 && msg.value != idsAndAmounts[0][1]) {
+                revert InvalidValue(msg.value, idsAndAmounts[0][1]);
+            }
+            idsAndAmounts[0][1] = msg.value;
+
+            idIndex++;
+        }
+
+        for (; idIndex < idsLength; idIndex++) {
+            (uint96 allocatorId, address token) = _splitId(idsAndAmounts[idIndex][0]);
+
+            // Check allocator id
+            if (allocatorId != ALLOCATOR_ID) {
+                revert InvalidAllocatorId(allocatorId, ALLOCATOR_ID);
+            }
+
+            if (idsAndAmounts[idIndex][1] == 0) {
+                // Amount is derived from the allocators token balance
+                idsAndAmounts[idIndex][1] = IERC20(token).balanceOf(address(this));
+            }
+
+            if (IERC20(token).allowance(address(this), address(_COMPACT)) < idsAndAmounts[idIndex][1]) {
+                SafeTransferLib.safeApproveWithRetry(token, address(_COMPACT), type(uint256).max);
+            }
+        }
+
+        return idsAndAmounts;
+    }
+
+    function _checkSignature(bytes32 digest, bytes calldata signature) internal view returns (bool) {
+        bytes32 r;
+        bytes32 s;
+        uint8 v;
+
+        if (signature.length == 65) {
+            (r, s) = abi.decode(signature, (bytes32, bytes32));
+            v = uint8(signature[64]);
+        } else if (signature.length == 64) {
+            bytes32 vs;
+            (r, vs) = abi.decode(signature, (bytes32, bytes32));
+            v = uint8(uint256(vs >> 255) + 27);
+            s = vs << 1 >> 1;
+        } else {
+            return false;
+        }
+
+        // Check if the signer is an authorized allocator address
+        return signers[ecrecover(digest, v, r, s)];
```

Comment:
for safety its often worth a special check and revert if ecrecover returns address(0) => ie invalid signature